feat: Upgrade Hive to 4.0.1 and align timestamp behaviour with SQL standard

imjalpreet opened the pull request:
## Description
<!---Describe your changes in detail-->
Upgrade to Hive 4.0.1 and align timestamp behaviour with SQL standard

Depends on https://github.com/prestodb/presto-hive-apache/pull/65, https://github.com/prestodb/tempto/pull/299, https://github.com/prestodb/orc-protobuf/pull/25, https://github.com/prestodb/presto-hive-dwrf/pull/13, https://github.com/prestodb/presto-hive-dwrf/pull/14 and https://github.com/prestodb/docker-images/pull/58

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->
#24435 and #27028

## Summary by Sourcery

Upgrade the Hive/ORC/Parquet stack and default timestamp semantics, wiring session-aware timezone handling through readers, writers, connectors, and tests.

New Features:
- Add support for writer and stripe timezones in ORC and Parquet metadata and readers, enabling standard-compliant timestamp handling.
- Expose connector session timestamp semantics to Hive, Cassandra, Iceberg, Hudi, and Parquet page sources and sinks for consistent behavior across connectors.

Bug Fixes:
- Align timestamp, date, and string statistics handling with upstream ORC/Parquet/Hive behavior to avoid incorrect predicate evaluation and metadata validation failures.
- Fix various tests and CI flows that depended on legacy timestamp semantics, incorrect timezones, or outdated expectations.

Enhancements:
- Upgrade Hive, ORC, DWRF, and Tempto dependencies and remove obsolete components such as hive-llap-common and PrestoAvroSerDe.
- Refine selective/batch readers and writers in ORC and Parquet to be timezone-aware and to honor microsecond-precision timestamps where applicable.
- Simplify S3 Select split schema handling by relying on LIST_COLUMNS/LIST_COLUMN_TYPES instead of Thrift DDL mangling.
- Improve materialized view and partition statistics handling by using session-aware partition value parsing and timestamp statistics computation.
- Adjust default function configuration to disable legacy timestamp semantics and update JSON/timestamp casts accordingly.

Build:
- Bump hive-apache, hive-dwrf, orc-protobuf, tempto, and related dependency versions and clean up transitive dependencies in Parquet, Iceberg, Pinot, and Hive function namespace modules.

CI:
- Relax product test workflow conditions to always run specific job groups when there are code changes, improving signal for storage-format and connector tests.

Tests:
- Update Hive, ORC, Parquet, Cassandra, Kafka, Iceberg, Druid, and expression engine tests to use the new timestamp semantics, UTC-based expectations, and session-aware timezone handling.
- Add Hive 4 insert-overwrite test coverage and adjust various smoke and integration tests to reflect upgraded Hive images and behavior.

Chores:
- Switch Docker-based Hive images to newer Hive 3/4 variants and adjust MinIO/HDFS wiring for tests and Presto-on-Spark environments.

## Sourcery Reviewer's Guide

https://github.com/prestodb/presto/pull/24571#issuecomment-4464222430

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->
Default value of `deprecated.legacy-timestamp` changed from `true` to `false`.

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [x] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Fix timestamp operations to match the SQL specification. The value of a TIMESTAMP type is not affected by the session time zone.
* Update default value of `deprecated.legacy-timestamp` to false.

Hive Connector Changes
* Upgrade to Hive 4.0.1
```

ScrapCodes commented:
Looking closely at the reason why tests are failing, beginning with presto-orc module.
Progress so far:
* Tests try to write content with hive and read by presto, especially in case of timestamps, it is incorrectly read by presto. Why exactly, I am still trying to understand - not clear yet.
The contents written by tests is correct because on reading with an external orc reader like my following python script, it gives correct result:
```py
from pyarrow import orc

table2 = orc.read_table('/tmp/3420396529049254202/data.orc')
print(table2)
```
```
python read_orc.py 
pyarrow.Table
test: timestamp[ns]
----
test: [[1970-01-01 00:00:00.001000000,1970-01-01 00:00:00.003000000,1970-01-01 00:00:00.005000000,1970-01-01 00:00:00.007000000,1970-01-01 00:00:00.011000000,...,1970-01-01 00:00:00.001000000,1970-01-01 00:00:00.003000000,1970-01-01 00:00:00.005000000,1970-01-01 00:00:00.007000000,1970-01-01 00:00:00.011000000]]

```

Whereas, record reader for ORC in presto reads it differently
![Screenshot from 2025-04-28 17-35-06](https://github.com/user-attachments/assets/d06191a9-4c6c-4c45-8d3a-8987fd127af0)

* These values e.g. `21600001` when converted to timestamp values give a 6h difference. e.g. 

![Screenshot from 2025-04-28 17-47-33](https://github.com/user-attachments/assets/18e90250-a5bc-4653-ba33-30297475b1a9)
This is the reason for failure of these tests.

ScrapCodes commented:
Another interesting find:
Trying to read directly by hive version 3.0.0, orc file produced by presto without hive 4 changes. Produces incorrect results, whereas it produces correct results when reading a file with hive 4 changes.

```
prashant@prashant:/drive1/work/temp/apache-hive-3.0.0-bin$ HADOOP_HOME=../hadoop-3.3.6/ bin/hive --orcfiledump ../../orc-reader/data-without-hive4.orc 

Processing data file ../../orc-reader/data-without-hive4.orc [length: 931]
Structure for ../../orc-reader/data-without-hive4.orc
File Version: 0.12 with ORC_135
Rows: 30000
Compression: ZLIB
Compression size: 262144
Type: struct<test:timestamp>

Stripe Statistics:
  Stripe 1:
    Column 0: count: 30000 hasNull: false
    Column 1: count: 30000 hasNull: false min: 1969-12-31 16:00:00.001 max: 1969-12-31 16:00:00.017

File Statistics:
  Column 0: count: 30000 hasNull: false
  Column 1: count: 30000 hasNull: false min: 1969-12-31 16:00:00.001 max: 1969-12-31 16:00:00.017

Stripes:
  Stripe: offset: 3 data: 342 rows: 30000 tail: 78 index: 379
    Stream: column 0 section ROW_INDEX start: 3 length 17
    Stream: column 1 section ROW_INDEX start: 20 length 61
    Stream: column 1 section BLOOM_FILTER start: 81 length 175
    Stream: column 1 section BLOOM_FILTER_UTF8 start: 256 length 126
    Stream: column 1 section DATA start: 382 length 23
    Stream: column 1 section SECONDARY start: 405 length 319
    Encoding column 0: DIRECT
    Encoding column 1: DIRECT_V2

File length: 931 bytes
Padding length: 0 bytes
Padding ratio: 0%
________________________________________________________________________________________________________________________


```

ScrapCodes commented:
The test failures such as:
`TestOrcReader>AbstractTestOrcReader.testLongDirect:158->AbstractTestOrcReader.testRoundTripNumeric:325 expected [1970-01-01 00:00:00.001] but found [1970-01-01 06:00:00.001]`

Is in the way presto reads the data and writes the data. Somehow even before the data is interpreted as a timestamp type i.e it is still a long type, it has a timestamp adjusted to system timezone. Why it happens is not yet clear to me, when the data written by presto is read via an external ORC reader it has a 6h adjustment applied to it. A similar thing happens when presto reads the data written by hive.

imjalpreet commented:
This is dependent on multiple PRs. First PR in line: https://github.com/prestodb/presto-hive-apache/pull/65

imjalpreet commented:
The following dependent PRs under review:

Tempto: https://github.com/prestodb/tempto/pull/299
orc-protobuf (for dwrf): https://github.com/prestodb/orc-protobuf/pull/25
presto-dwrf: https://github.com/prestodb/presto-hive-dwrf/pull/13 (JDK Upgrade) followed by https://github.com/prestodb/presto-hive-dwrf/pull/14 (Hive Upgrade)

imjalpreet commented:
Cassandra Product Tests failing due to https://github.com/prestodb/tempto/pull/298

imjalpreet commented:
Need to debug why a couple of recently added native sidecar plugin unit tests are failing after the latest rebase

imjalpreet commented:
> Need to debug why a couple of recently added native sidecar plugin unit tests are failing after the latest rebase

```
TestNativeSidecarPlugin.testNativeExpressionOptimizer:828->getNowInstant:901->AbstractTestQueryFramework.computeActual:149 ? Runtime  Timezone cannot be null native.default.fail(Timezone cannot be null:VARCHAR) Top-level Expression: cast((native.default.fail(Timezone cannot be null:VARCHAR)) as TIMESTAMP WITH TIME ZONE)
```

Found a bug in Velox due to which a couple of sidecar tests were failing. If  `adjustTimestampToTimezone` is false, timezone returns a nullptr, and it's not handled properly in Velox.

https://github.com/facebookincubator/velox/blob/805db6b72fe326e7b2c97666fe4ba04a7e448568/velox/expression/CastExpr.cpp#L41-L49
+
https://github.com/facebookincubator/velox/blob/805db6b72fe326e7b2c97666fe4ba04a7e448568/velox/functions/prestosql/DateTimeFunctions.h#L1835-L1838

I will be raising a bug in Velox for this.

imjalpreet commented:
@sourcery-ai summary

imjalpreet commented:
@sourcery-ai guide

imjalpreet committed:
feat: Upgrade Hive to 4.0.1 and align timestamps with SQL standard

imjalpreet commented:
PR is waiting for a review.

nmahadevuni commented on presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestArrowFlightQueries.java:
```diff
@@ -131,11 +131,12 @@ public void testSelectTime()
                         getDate("2004-12-31"),
-                        getTimeAtZone("23:59:59", session.getTimeZoneKey()),
-                        getDateTimeAtZone("2005-12-31 23:59:59", session.getTimeZoneKey()))
+                        getTimeAtZone("23:59:59", timeZoneKey),
```
Let's run the test with both legacy and UTK_KEY?

nmahadevuni commented on presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPageSink.java:
```diff
@@ -156,9 +162,12 @@ else if (REAL.equals(type)) {
             values.add(toCassandraDate.apply(type.getLong(block, position)));
         }
-        else if (TIMESTAMP.equals(type)) {
+        else if (session.getSqlFunctionProperties().isLegacyTimestamp() && TIMESTAMP.equals(type)) {
```
Is this mutually exclusive? If its legacy, then it must be only TIMESTAMP without TIME ZONE?

nmahadevuni replied:
Please add a comment why its so for cassandra.

imjalpreet replied:
Yes, it's mutually exclusive. The implementation before this PR will be treated as legacy behaviour(not aligned with Cassandra Timestamp Semantics) and, for backwards compatibility, is mapped to timestamp, and the new implementation after this PR is aligned with Cassandra UTC semantics and is mapped to TIMESTAMP_WITH_TIME_ZONE.

I have added code comments for clarity

nmahadevuni commented on presto-cassandra/src/test/java/com/facebook/presto/cassandra/TestCassandraIntegrationSmokeTest.java:
```diff
@@ -66,6 +68,7 @@ public class TestCassandraIntegrationSmokeTest
 
+    private static final ZonedDateTime TIMESTAMP_VALUE = ZonedDateTime.of(1970, 1, 1, 3, 4, 5, 0, ZoneId.of("UTC"));
     private static final Timestamp DATE_TIME_LOCAL = Timestamp.valueOf(LocalDateTime.of(1970, 1, 1, 3, 4, 5, 0));
     private static final LocalDateTime TIMESTAMP_LOCAL = LocalDateTime.of(1969, 12, 31, 23, 4, 5); // TODO #7122 should match DATE_TIME_LOCAL
```
This is unused. Can remove it.

nmahadevuni replied:
Same for DATE_TIME_LOCAL

nmahadevuni commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/TimestampPlainValuesDecoder.java:
```diff
@@ -48,7 +52,9 @@ public void readNext(long[] values, int offset, int length)
 
         while (offset < endOffset) {
-            values[offset++] = getTimestampMillis(localByteBuffer, localBufferOffset);
+            AtomicLong utcMillis = new AtomicLong(getTimestampMillis(localByteBuffer, localBufferOffset));
```
why do we need AtomicLong here?

imjalpreet replied:
We actually don't need it and can be refactored. It was used only because lambda allows only effectively final variables, and, probably in some iteration of the PR, this change was made. I have refactored to remove the usage of lambda and updated the variable to a simple long.

nmahadevuni commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/Int64TimeAndTimestampMicrosNestedBatchReader.java:
Don't we need to apply the timezone conversion for this?

imjalpreet replied:
`Int64TimeAndTimestampMicrosValuesDecoder` handles timezone at construction time (via the withTimezone flag) rather than at readNext time, so timezone is not forwarded here.

https://github.com/prestodb/presto/blob/master/presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/Int64TimeAndTimestampMicrosPlainValuesDecoder.java#L43-L49

nmahadevuni commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/Int64TimeAndTimestampMicrosNestedBatchReader.java:
```diff
@@ -90,7 +91,7 @@ else if (definitionLevels[definitionLevelIndex] == maxDefinitionLevel - 1) {
 
     @Override
-    protected ColumnChunk readNestedNoNull()
+    protected ColumnChunk readNestedNoNull(Optional<DateTimeZone> timezone)
```
Same comment as in above method.

imjalpreet replied:
Same as above

nmahadevuni commented on presto-parquet/src/main/java/com/facebook/presto/parquet/reader/BinaryColumnReader.java:
```diff
@@ -37,7 +40,7 @@ public BinaryColumnReader(RichColumnDescriptor descriptor)
 
     @Override
-    protected void readValue(BlockBuilder blockBuilder, Type type)
+    protected void readValue(BlockBuilder blockBuilder, Type type, Optional<DateTimeZone> timezone)
```
We are making changes to all non-timestamp readers methods because we want to pass the DateTimeZone. Can we make an abstract class for Timestamp related class and derive all Timestamp readers from it to contain this to few related files?

imjalpreet replied:
Yes, that makes sense. I will refactor.

nmahadevuni commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -598,21 +600,13 @@ private static void initializeDeserializer(Configuration configuration, Deserial
             configuration = copy(configuration); // Some SerDes (e.g. Avro) modify passed configuration
-            deserializer.initialize(configuration, schema);
-            validate(deserializer);
+            ((AbstractSerDe) deserializer).initialize(configuration, schema, null);
```
why don't we need to validate the deserializer anymore?

imjalpreet replied:
This is not required anymore as Hive 4.x changed `AbstractSerDe.initialize()` to throw `SerDeException` directly on configuration errors. The separate `getConfigurationErrors()` check from Hive 3.x was removed from the API. I have handled errors via try-catch now due to this change.

nmahadevuni commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java:
```diff
@@ -339,10 +347,16 @@ static StringStatistics toStringStatistics(HiveWriterVersion hiveWriterVersion,
-        Slice maximum = stringStatistics.hasMaximum() ? maxStringTruncateToValidRange(byteStringToSlice(stringStatistics.getMaximumBytes()), hiveWriterVersion) : null;
-        Slice minimum = stringStatistics.hasMinimum() ? minStringTruncateToValidRange(byteStringToSlice(stringStatistics.getMinimumBytes()), hiveWriterVersion) : null;
+        Slice maximum = stringStatistics.hasUpperBound() ?
+                maxStringTruncateToValidRange(byteStringToSlice(stringStatistics.getUpperBoundBytes()), hiveWriterVersion) :
```
these changes look unrelated, why are they here?

imjalpreet replied:
No, these changes have come as part of the Hive 4.x upgrade. There are certain statistics changes.

imjalpreet replied:
ORC changed how it handles StringStatistics (see [PR](https://github.com/apache/orc/pull/292/files))

nmahadevuni reviewed (commented):
Thanks @imjalpreet . Reviewed a part of the code. Can you add new or modify current tests to test the new timestamp behaviour as well as old?

nmahadevuni commented on presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java:
```diff
@@ -187,7 +184,7 @@ public void testTimestampLiteralPushdown()
                         ImmutableList.of("city", "datetime")),
-                "SELECT \"city\", \"datetime\" FROM \"realtimeOnly\" WHERE (\"datetime\" = TIMESTAMP '2016-06-26 06:00:00.000')");
+                "SELECT \"city\", \"datetime\" FROM \"realtimeOnly\" WHERE (\"datetime\" = TIMESTAMP '2016-06-26 19:00:00.000')");
         //test timestamp with timezone
```
The result of this test seems to be the same as the next test. Should we test with legacy timestamp for this one?

imjalpreet replied:
Yes, but this one is for Timestamp, and the next one is for Timestamp with Timezone. They exercise different code branches in DruidPushdownUtils.getConstantAsString

nmahadevuni commented on presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchQueryBuilder.java:
```diff
@@ -161,10 +162,10 @@ private static FieldValue getValue(ConnectorSession session, Type type, Object v
-
             String dateValue = Instant.ofEpochMilli((Long) value)
-                    .atZone(ZoneId.of(session.getSqlFunctionProperties().getTimeZoneKey().getId()))
+                    .atZone(session.getSqlFunctionProperties().isLegacyTimestamp() ?
```
Are there tests to verify with both legacy and new behaviour?

imjalpreet replied:
There are timestamp tests executed via TestElasticsearchIntegrationSmokeTest (which extends AbstractTestIntegrationSmokeTest and AbstractTestQueryFramework). These will test the new default behaviour.

For legacy behaviour, we can add tests if needed via a follow-up PR.

imjalpreet replied:
I will add a UT for this PR

nmahadevuni commented on presto-hive/src/main/java/com/facebook/presto/hive/util/SerDeUtils.java:
```diff
@@ -279,35 +278,27 @@ private static long formatDateAsLong(Object object, DateObjectInspector inspecto
     }
 
-    private static long formatTimestampAsLong(Object object, TimestampObjectInspector inspector, DateTimeZone hiveStorageTimeZone)
+    private static long formatTimestampAsLong(Object object, TimestampObjectInspector inspector, DateTimeZone hiveStorageTimeZone, boolean legacyTimestampEnabled)
```
hiveStorageTimeZone already has UTC or a different timezone, do we need to pass legacyTimestampEnabled too? just `hiveStorageTimeZone.convertLocalToUTC(hiveMillis, false) ` should be enough? which would be a no-op if timezone is already UTC?

imjalpreet replied:
No, we still need the legacy timestamp check for two reasons. First, it keeps the behavior consistent with other parts of Presto. Second, we cannot assume that users will configure hiveStorageTimezone to UTC. For backward compatibility, there is currently no restriction preventing the storage time zone from being set to a non-UTC value, even when legacy timestamps are disabled by default. In such cases, the timezone shift should be applied only when the legacy timestamp mode is enabled through the session property, otherwise, no timezone shift should be performed.

Also,  to keep backward compatibility, for certain file formats where we are reading the timestamp from the file footer, we allow users to configure a non-UTC timezone(even when legacy timestamp is disabled) for older files where the footer might not contain the timezone information.

nmahadevuni commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -471,9 +471,11 @@ public static long parseHiveDate(String value)
+    public static long parseHiveTimestamp(ConnectorSession session, String value, DateTimeZone timeZone)
     {
-        return HIVE_TIMESTAMP_PARSER.withZone(timeZone).parseMillis(value);
+        return session != null && session.getSqlFunctionProperties().isLegacyTimestamp() ?
```
pass just isLegacyTimestamp from the caller making the check there?

nmahadevuni commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastoreClient.java:
```diff
@@ -248,7 +253,8 @@ public void setTableColumnStatistics(String databaseName, String tableName, List
     {
-        client.delete_table_column_statistics(constructSchemaName(catalogName, databaseName), tableName, columnName);
+        // TODO: This is not backwards compatible
+        client.delete_table_column_statistics(constructSchemaName(catalogName, databaseName), tableName, columnName, "hive");
```
what's the change that requires hardcoded argument "hive" to be sent to the metastore? Can we create a static final member for this?

imjalpreet replied:
The Hive 4.x Thrift API added a engine parameter to both delete_*_column_statistics calls, and "hive" is the engine name for the default Hive statistics engine. Earlier it used to be fixed as hive but now thrift API allows each engine to maintain its own version of column stats. For backwards compatibility, we are leaving it as hive.

nmahadevuni commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java:
```diff
@@ -2137,12 +2137,10 @@ private void testBucketTransformsOnTimeForFormat(Session session, FileFormat for
         assertUpdate(session, insertSql, 7);
 
-        assertQuery(session, "SELECT COUNT(*) FROM \"test_bucket_transform_on_time$partitions\"", "SELECT 4");
+        assertQuery(session, "SELECT COUNT(*) FROM \"test_bucket_transform_on_time$partitions\"", "SELECT 2");
```
Why does this change lead to different number of partitions?

imjalpreet replied:
Iceberg's bucket transform hashes the internal numeric encoding of the value. For TIME, that encoding is microseconds since midnight. 

I debugged the test again, and below are my findings.

Before this PR (`legacy` mode): `TIME` values were encoded with the session timezone offset applied, so for eg time '01:02:03.123' in a session with timezone `America/Bahia_Banderas` (UTC-7, the test default) would have a different microsecond value than in UTC. That offset shifted the values before hashing, distributing them across all 4 buckets.

After this PR (non-legacy, SQL standard): `TIME` values are encoded as pure microseconds since midnight UTC, no session timezone adjustment. The same 7 time literals now hash with their true microsecond values, and it happens that all 7 fall into only 2 of the 4 buckets (0 and 2).

So the partition count drop from 4 to 2 is an expected consequence of fixing the TIME encoding, not a regression.

nmahadevuni commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergFileWriterFactory.java:
```diff
@@ -94,14 +97,16 @@ public IcebergFileWriterFactory(
         this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion is null");
         this.orcFileWriterConfig = requireNonNull(orcFileWriterConfig, "orcFileWriterConfig is null");
         this.dwrfEncryptionProvider = requireNonNull(dwrfEncryptionProvider, "DwrfEncryptionProvider is null").toDwrfEncryptionProvider();
+        this.writerTimezone = requireNonNull(hiveClientConfig, "hiveClientConfig is null").getDateTimeZone();
```
Don't we apply the same legacy vs new semantics for Iceberg?

imjalpreet replied:
This behaviour is consistent with Hive connector's com.facebook.presto.hive.parquet.ParquetFileWriterFactory.

nmahadevuni commented on presto-main-base/src/main/java/com/facebook/presto/util/JsonUtil.java:
```diff
@@ -544,7 +545,7 @@ public void writeJsonValue(JsonGenerator jsonGenerator, Block block, int positio
             else {
                 long value = TIMESTAMP.getLong(block, position);
-                jsonGenerator.writeString(printTimestampWithoutTimeZone(properties.getTimeZoneKey(), value));
+                jsonGenerator.writeString(printTimestampWithoutTimeZone(properties.isLegacyTimestamp() ? properties.getTimeZoneKey() : UTC_KEY, value));
```
We could do the same in JsonOperators.java instead of branching on isLegacyTimestamp.

imjalpreet commented:
Currently, working on the rebase with master to make Cassandra 4.x upgrade compatible with the timestamp semantics changes.

imjalpreet committed:
fix(plugin-cassandra): Use UTC instant for smoke test data setup

Timestamp.valueOf(LocalDateTime) interprets the local time in the JVM default timezone (America/Bahia_Banderas = UTC-7 during tests), causing the stored Cassandra instant to be 7 hours ahead of what the test assertions expected. Replace with an explicit ZoneOffset. UTC instant so the stored value matches the expected 1970-01-01T03:04:05Z regardless of JVM timezone. Also remove unused DATE_TIME_LOCAL / TIMESTAMP_LOCAL fields and unused imports from the test class.

tdcmeehan commented on .github/workflows/product-tests-specific-environment.yml:
```diff
@@ -77,25 +77,25 @@ jobs:
         run: presto-product-tests/bin/run_on_docker.sh singlenode -g hdfs_no_impersonation,avro,mixed_case
       - name: Product Tests Specific 1.2
-        if: needs.changes.outputs.codechange == 'true'
+        if: needs.changes.outputs.codechange == 'true' && ${{ always() }}
```
Why do we need this change?

imjalpreet replied:
@tdcmeehan It’s not related to this PR, so I can raise it in a separate PR if needed. 

The reason for this change is that this module includes seven product test suites, all of which are independent of each other. Currently, they run sequentially, and if one suite fails, the entire test run fails immediately, preventing the remaining suites from running.

I think it would be better to allow all the suites to run since they’re unrelated. Otherwise, if multiple suites have failures, we end up fixing one issue, waiting for the entire suite to run again, and only then discovering that another suite has also failed.

tdcmeehan commented on presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordCursor.java:
```diff
@@ -16,27 +16,33 @@
         implements RecordCursor
 {
     private final List<FullCassandraType> fullCassandraTypes;
+    private ConnectorSession session;
```
```suggestion
    private final ConnectorSession session;
```

tdcmeehan reviewed (dismissed):
Just a couple of nits

imjalpreet committed:
Address review comments from @nmahadevuni

imjalpreet commented:
Both test failures are unrelated.

- [arrow flight tests / arrowflight-native-e2e-tests (:presto-flight-shim) (pull_request)](https://github.com/prestodb/presto/actions/runs/29855233997/job/88914362427?pr=24571) is flaky and I have raised an issue here: https://github.com/prestodb/presto/issues/28205
- [codenotify / Label and notify (pull_request_target)](https://github.com/prestodb/presto/actions/runs/29855232411/job/88717789524?pr=24571) should be fixed after https://github.com/prestodb/presto/pull/28198 is merged
